### ISSUE TITLE
Feature Request: CN/UPX

### ISSUE THREAD

**Issue by Lonnegan** [xmrig/xmrig#2274]
Hi,

is there the possibility to support CN/UPX officially? The coin goes thru the roof atm and there is only an old non-opensource variant of xmrig available for UPX. But I know, that you can do this miles better :)

**Comment by SChernykh** [comment]
https://github.com/xmrig/xmrig/pull/2276
https://github.com/xmrig/xmrig-cuda/pull/95

**Comment by SChernykh** [comment]
Bonus: MSVC version for Windows is faster than other miners by 3-4% on Ryzen CPUs.

**Comment by SChernykh** [comment]
Update: https://github.com/xmrig/xmrig/pull/2278 - GCC version is a bit faster than MSVC now. This code is now the fastest on Zen3 with big margin (**7.8%** faster than SRBMiner, **15.7%** faster than XMRigCC on my **Ryzen 5 5600X**).

**Comment by Lonnegan** [comment]
I am looking forward to the release so I can test it myself :)

**Comment by Lonnegan** [comment]
I see you've made an extra asm for Zen3. Is the reason just because you have one ;) or are there specific things on Zen3 to make the effort?

And a second question: is UPX really a variant of cn/lite? Im asking this because cn/lite years ago worked very well on week CPUs with big caches (like AMD Bulldozer) doing Dual-Share or low-power-mode or what ever it was called in the past. With a scratchpad size of just 128 KB there would fit multiples (16) shares into the L2 cache of one Bulldozer module! :o :D

**Comment by SChernykh** [comment]
No, I just discovered that compiler generated code runs faster on Zen3 so I took it as is. I haven't started optimizing it specifically for Zen3 yet. UPX is a variant of Cryptonight V2 with smaller scratchpad and fewer iterations.

**Comment by Spudz76** [comment]
It is referred to as a cn/femto so yes it's like cn/lite except even smaller by half again.  Should work nice for low-cache but fast-core.

**Comment by xmrig** [comment]
[v6.12.0](https://github.com/xmrig/xmrig/releases/tag/v6.12.0) released.

**Comment by Shai0Hulud** [comment]
> Update: #2278 - GCC version is a bit faster than MSVC now. This code is now the fastest on Zen3 with big margin (**7.8%** faster than SRBMiner, **15.7%** faster than XMRigCC on my **Ryzen 5 5600X**).

Ha, I was hoping it's also faster on Zen 2^^

GCC is exactly as fast as XMRigCC. Which is more than okay.

Thanx for all your work guys!

**Comment by SChernykh** [comment]
> You mentioned zen3 option, I guess for asm option? Is there documentation what options are possible? Is there one for Ryzen Zen 2 CPUs for example?

`"asm":true,` in config.json should be enough to automatically detect the best code for your CPU. Currently there's a dedicated code for Zen3 and a single version for all other CPUs.

**Comment by benthetechguy** [comment]
> there is only an old non-opensource variant of xmrig available for UPX

XMRig is licensed under GPLv3. Failure to provide source code for a fork of XMRig is a violation of its license.
